### ISSUE TITLE
jqjq support

### ISSUE THREAD

**Issue by Freed-Wu** [01mf02/jaq#216]
https://github.com/wader/jqjq needs `-L`

**Comment by wader** [comment]
It would be fun if jaq could run jqjq 😄 some months ago i hacked around the lack of -L support (but it seem to have it now?) but then ran into lack of destructing support.
You can test it like this if you want:
```sh
$ jaq -n -L . 'include "jqjq"; eval("1+2")'
Error: expected variable
   ╭─[/Users/wader/src/jqjq/jqjq.jq]
   │
49 │         | ([.[2:6], .[8:] | _fromhex]) as [$hi,$lo]
   ┆                                           ────┬────
   ┆                                               │
   ┆                                               ╰───── unexpected token
...
```

**Comment by 01mf02** [comment]
`-L` will be supported in jaq 2.0.

And yes, I would love to see jaq being able to run jqjq, but lack of destructuring support is indeed a pretty large road-block. Perhaps in jaq 3.0? :)

**Comment by wader** [comment]
> `-L` will be supported in jaq 2.0.

Ah yes i probably tested with current main branch

> And yes, I would love to see jaq being able to run jqjq

Would be great and a fun project to help out with

> but lack of destructuring support is indeed a pretty large road-block. Perhaps in jaq 3.0? :)

One of my favorite but sadly a bit hidden features of jq! sometimes i also dream about how some kind of pattern matching could work in jq 🤔

**Comment by 01mf02** [comment]
@wader, I was actually able to implement destructuring after all in jaq relatively quickly. :)
My current prototype is in the [`patterns`](https://github.com/01mf02/jaq/tree/patterns) branch, if you wish to try it.
I'll try running jqjq with it later myself.

**Comment by wader** [comment]
🥳  nice, i gave it a try. with this jqjq changes i can get it to lex but parsing seems to bail out in the precedence climbing
code, digging into that.

some notes:
- `match` with `^...`and `m` flag seem to match any line or something in jaq but not jq.
- `{` not need to be escaped in jq regexp but in jaq
- `gamma`, `getpath`, `setpath` and `delpaths` not available in jaq

```patch
diff --git a/jqjq.jq b/jqjq.jq
index bc5e632..64d0fde 100644
--- a/jqjq.jq
+++ b/jqjq.jq
@@ -80,7 +80,7 @@ def lex:
     def _re($re; f):
       ( . as {$remain, $string_stack}
       | $remain
-      | match($re; "m").string
+      | match($re; "").string
       | f as $token
       | { result: ($token | del(.string_stack))
         , remain: $remain[length:]
@@ -153,8 +153,8 @@ def lex:
       // _re("^\\)";    {rparen: ., string_stack: ($string_stack[0:-1])})
       // _re("^\\[";    {lsquare: .})
       // _re("^\\]";    {rsquare: .})
-      // _re("^{";      {lcurly: .})
-      // _re("^}";      {rcurly: .})
+      // _re("^\\{";    {lcurly: .})
+      // _re("^\\}";    {rcurly: .})
       // _re("^\\.\\."; {dotdot: .})
       // _re("^\\.";    {dot: .})
       // _re("^\\?";    {qmark: .})
@@ -228,7 +228,7 @@ def parse:
         else false
         end;
 
-      ( _p("query1") as [$rest, $t]
+      ( debug({prec: .}) | _p("query1") as [$rest, $t]
       | $rest
       | def _f($t):
           ( .[0] as $next # peek next
@@ -272,6 +272,7 @@ def parse:
     def _scalar($type; c; f):
       ( . as [$first]
       | _consume(c)
+      | debug({aaa:.})
       | [ .
         , { term:
               ( $first
@@ -280,6 +281,7 @@ def parse:
               )
           }
         ]
+      | debug({bbb:.})
       );
 
     # {<keyval>...} where keyval is:
@@ -1060,7 +1062,7 @@ def parse:
         ]
       );
 
-    ( .# debug({_p: $type})
+    ( debug({_p: $type})
     | if $type == "query" then
         _op_prec_climb(0; false)
       elif $type == "keyval_query" then
@@ -1112,7 +1114,7 @@ def parse:
           // _p("recurse") # ".."
           ) as [$rest, $term]
         | $rest
-        | _repeat(_p("suffix")) as [$rest, $suffix_list]
+        | _repeat(empty | _p("suffix")) as [$rest, $suffix_list]
         | $rest
         | [ .
           , ( $term
@@ -1516,10 +1518,10 @@ def eval_ast($query; $path; $env; undefined_func):
                 ( a0 as $a0
                 | [[null], has($a0)]
                 )
-              elif $name == "delpaths/1" then
-                ( a0 as $a0
-                | [[null], delpaths($a0)]
-                )
+              # elif $name == "delpaths/1" then
+              #   ( a0 as $a0
+              #   | [[null], delpaths($a0)]
+              #   )
               elif $name == "explode/0"  then [[null], explode]
               elif $name == "implode/0"  then [[null], implode]
               elif $name == "tonumber/0" then [[null], tonumber]
@@ -1536,19 +1538,19 @@ def eval_ast($query; $path; $env; undefined_func):
                 | error($a0)
                 )
               elif $name == "halt_error/1" then [[null], halt_error(a0)]
-              elif $name == "getpath/1" then
-                ( a0 as $a0
-                | [ $path+$a0
-                  , getpath($a0)
-                  ]
-                )
-              elif $name == "setpath/2" then
-                ( a0 as $a0
-                | a1 as $a1
-                | [ []
-                  , setpath($a0; $a1)
-                  ]
-                )
+              # elif $name == "getpath/1" then
+              #   ( a0 as $a0
+              #   | [ $path+$a0
+              #     , getpath($a0)
+              #     ]
+              #   )
+              # elif $name == "setpath/2" then
+              #   ( a0 as $a0
+              #   | a1 as $a1
+              #   | [ []
+              #     , setpath($a0; $a1)
+              #     ]
+              #   )
               elif $name == "path/1" then
                 ( _e($args[0]; []; $query_env) as [$p, $_v]
                 # TODO: try/catch error
@@ -1577,7 +1579,7 @@ def eval_ast($query; $path; $env; undefined_func):
               elif $name == "expm1/0"       then [[null], expm1]
               elif $name == "fabs/0"        then [[null], fabs]
               elif $name == "floor/0"       then [[null], floor]
-              elif $name == "gamma/0"       then [[null], gamma]
+              # elif $name == "gamma/0"       then [[null], gamma]
               elif $name == "j0/0"          then [[null], j0]
               elif $name == "j1/0"          then [[null], j1]
               elif $name == "lgamma/0"      then [[null], lgamma]
@@ -2538,9 +2540,10 @@ def eval($expr; $globals; $builtins_env):
   # TODO: does not work with jq yet because issue with bind patterns
   # $ gojq -cn -L . 'include "jqjq"; {} | {a:1} | eval(".a") += 1'
   # {"a":2}
-  | if $path | . == [] or . == [null] then $value
-    else getpath($path)
-    end
+  # | if $path | . == [] or . == [null] then $value
+  #   else getpath($path)
+  #   end
+  | $value
   );
 def eval($expr):
   eval($expr; {}; _builtins_env);

```

**Comment by pkoppstein** [comment]
@wader wrote:
> match with ^...and m flag seem to match any line or something in jaq but not jq.

Just a reminder that jq's actual behavior regarding some of the regex
flags is generally not a good guide. Some details are at
https://github.com/jqlang/jq/issues/2663

@01mf02 -- I noticed that in your revision of the jq manual,
you added a "Compatibility" note regarding jaq and the regex library.
I was thinking that if you come across any instances where jaq's handling of regex flags is correct
but differs from that of jq 1.7, it would be very helpful to add some details.

**Comment by wader** [comment]
@pkoppstein true! haven't dugg any deeper what going on, here is a repro of the difference. seem to only happen when there is a `\n\n`

```sh
$ jq 'match("^\\s+"; "m")' <<< '"a\nb\n"'
$ jaq 'match("^\\s+"; "m")' <<< '"a\nb\n"'

$ jq 'match("^\\s+"; "m")' <<< '"a\n\nb\n"'
$ jaq 'match("^\\s+"; "m")' <<< '"a\n\nb\n"'
{
  "offset": 2,
  "length": 1,
  "string": "\n",
  "captures": []
}

```

**Comment by 01mf02** [comment]
> 🥳 nice, i gave it a try. with this jqjq changes i can get it to lex but parsing seems to bail out in the precedence climbing
> code, digging into that.

Great. Thanks for trying this out. Let us know if you need any help.

> @pkoppstein true! haven't dugg any deeper what going on, here is a repro of the difference. seem to only happen when there is a `\n\n`

jaq enables regex's [`multi_line`](https://docs.rs/regex/latest/regex/struct.RegexBuilder.html#method.multi_line) function when `m` is given. The behaviour that we see here is that `^` matches the `\n` (by definition), and because the `\n` just after it is a space character, the regex matches.

> @01mf02 -- I noticed that in your revision of the jq manual,
> you added a "Compatibility" note regarding jaq and the regex library.
> I was thinking that if you come across any instances where jaq's handling of regex flags is correct
> but differs from that of jq 1.7, it would be very helpful to add some details.

I'm not sure whether I'm the right person to do this --- I'm not very fluent at regexes, and even when I come across a discrepancy between jq and jaq, I am not confident to decide whether jq's or jaq's behaviour is right.
I also think that this might be a rabbit-hole from which one will have a hard time getting out. Because I'm sure that there are tons of differences between Oniguruma and regex, and I think that the jq manual is not a good point to document these differences.

**Comment by pkoppstein** [comment]
@01mf02 wrote:
> this might be a rabbit-hole 

Yes, but I think we could avoid that by focusing only on those cases where someone notices a discrepancy where (a) jaq and/or gojq is correct and (b) either jq 1.7 behavior is definitely wrong, or the jq manual is definitely in need of clarification in light of the correct behavior.

I would certainly be glad to help, and suspect @wader would too :-)

**Comment by 01mf02** [comment]
@pkoppstein, thanks for offering your help. I will first await the merging of the current state of the manual, before augmenting it with more content. So let us discuss the topic of regex documentation once this is done.

**Comment by pkoppstein** [comment]
@01mf02 - I'm sorry I probably can't do much to help expedite the acceptance of your excellent work on the manual.  Let me know if you think otherwise.

**Comment by wader** [comment]
Spent some more time on jqjq support and think i found a bug/difference between jaq and jq when it comes to destructing. Managed to minimize it down to this:

```sh
$ jq -cn '[{a:123}] | debug | .[0] as {$a} | debug, $a'
["DEBUG:",[{"a":123}]]
["DEBUG:",[{"a":123}]]
[{"a":123}]
123

$ jaq -cn '[{a:123}] | debug | .[0] as {$a} | debug, $a'
["DEBUG:", [{"a":123}]]
["DEBUG:", {"a":123}]
{"a":123}
123
```
It seems like the as-binding does not passthru the input unchanged in the case of `.[<index>] as {...} or [...]`

**Comment by wader** [comment]
Ran into this also, i would guess related to same issue:
```sh
$ jq -n '{} as $a | $a as {$t} | debug'
["DEBUG:",null]
null

$ jaq -n '{} as $a | $a as {$t} | debug'
["DEBUG:", {}]
{}
```

**Comment by 01mf02** [comment]
Hi @wader, just as I was setting out to look into jqjq myself right now, I find that you have beaten me to it. :)
Thanks for uncovering this bug. I have corrected binding behaviour in 4560c86.

**Comment by 01mf02** [comment]
Perhaps it is interesting to mention at this point that jaq's destructuring follows the same rules as its indexing, which diverge from jq by failing to index any null value. That means that `[] as [{$x}] | $x` yields `null` in jq, but an error in jaq, because `[] | .[0] | .x` is a failure too. However, `[] as [$x] | $x` yields `null` for both jq and jaq.

**Comment by wader** [comment]
Yeap noticed :) have done work workaround for it. I pushed my current progress here https://github.com/wader/jqjq/commits/jaq/ but be aware of lots of hacks :) with this at least "1+2" seem to kind of work but other expressions fails.

**Comment by wader** [comment]
> Hi @wader, just as I was setting out to look into jqjq myself right now, I find that you have beaten me to it. :) Thanks for uncovering this bug. I have corrected binding behaviour in [4560c86](https://github.com/01mf02/jaq/commit/4560c86eb09ca2211c76706f02033464edd35ae2).

That was quick! 🥳 let me know how it goes, and i will probably continue later today also

**Comment by 01mf02** [comment]
Thanks for mentioning the jaq branch of jqjq.
I found that quite a few things already work, using:

    jaq -n -L . 'include "jqjq"; eval($FILTER; {}; {})

This evaluates the filter without using any globals or builtins.

For $FILTER, the following work and yield the expected output:

* `1+2*3` --- yay, precedences
* `1.4e2`
* `[1, 2]`
* `{a: 1, b: (2, 3)}`
* `def f: 1; f`
* `if true then 0 else 1 end`
* `reduce (1, 2) as $x (0; .+$x)`
* `1 | .+2`

What does not work:

`.`, `.[]`, `.[0]`, `.a`, `1 as $x | $x`: These fail with: "cannot use null as iterable (array or object)".
For some of these, parsing seems to fail (`.`), for some others, execution.

**Comment by 01mf02** [comment]
Concerning the lexing speed:

    jaq -L . -n 'include "jqjq"; _builtins_src | lex'

is indeed slower than jq right now (1.2 seconds in jaq vs. 0.4 seconds in jq).
If it is much slower than that for you: Did you compile jaq with `cargo build --release`?

I just made some small experiment, and the performance problem seems to be related to slow regex execution. So I swapped the underlying regex library in 0840b1b010fd230be0c29011fd12654a75ce9363 to `regex-lite`, and this brings down the execution time in jaq to 0.4 seconds as well. :)

**Comment by wader** [comment]
> Thanks for mentioning the jaq branch of jqjq. I found that quite a few things already work, using:
> 
> ```
> jaq -n -L . 'include "jqjq"; eval($FILTER; {}; {})
> ```
> 
> This evaluates the filter without using any globals or builtins.

Hmm it seem to get stuck with `eval/1`, will have to be investigated.

> For $FILTER, the following work and yield the expected output:
> 
> * `1+2*3` --- yay, precedences
> * `1.4e2`
> * `[1, 2]`
> * `{a: 1, b: (2, 3)}`
> * `def f: 1; f`
> * `if true then 0 else 1 end`
> * `reduce (1, 2) as $x (0; .+$x)`
> * `1 | .+2`

🥳 

> What does not work:
> 
> `.`, `.[]`, `.[0]`, `.a`, `1 as $x | $x`: These fail with: "cannot use null as iterable (array or object)". For some of these, parsing seems to fail (`.`), for some others, execution.

I guess some null checks etc is needed in some places, i've neem bit sloppy with using null it seems :)

**Comment by wader** [comment]
> Concerning the lexing speed:
> 
> ```
> jaq -L . -n 'include "jqjq"; _builtins_src | lex'
> ```
> 
> is indeed slower than jq right now (1.2 seconds in jaq vs. 0.4 seconds in jq). If it is much slower than that for you: Did you compile jaq with `cargo build --release`?

Doh, yes you correct, i build without `--release`. Now i've updated to latest main and built with release target, much faster 👍

> I just made some small experiment, and the performance problem seems to be related to slow regex execution. So I swapped the underlying regex library in [0840b1b](https://github.com/01mf02/jaq/commit/0840b1b010fd230be0c29011fd12654a75ce9363) to `regex-lite`, and this brings down the execution time in jaq to 0.4 seconds as well. :)

When i read about regex-list it seem to focus on binary size, compile time and some less features (less unicode support?) over performance? but it was faster in this case also? ... oh they mean compile time as in time compiling a regex not rust build time? :)

BTW for regex heavy filters maybe a regex cache would be an interesting idea? there has been some talks in jq issues about various ways of doing, ex in https://github.com/jqlang/jq/issues/2856, maybe a LRU cache etc? i've also experimented a bit with how `/regex/` literals could work in jq, that would be neat and also simplify how to reuse a compiled regex.

**Comment by wader** [comment]
Any idea what is going on with these? https://github.com/wader/jqjq/blob/master/jqjq.jq#L1945-L1946

Without any changes it seem expressions like  `1+2` just output `1`, but if i add `... | debug` or `... | .` to both it seem to work. Some tail or return optimisation going wrong? 🤔

**Comment by wader** [comment]
> What does not work:
> 
> `.`, `.[]`, `.[0]`, `.a`, `1 as $x | $x`: These fail with: "cannot use null as iterable (array or object)". For some of these, parsing seems to fail (`.`), for some others, execution.

Now all of those work in the jaq branch, and i think all of them were different variants of null index. Current fixes are mostly temp workarounds, should try figure something nicer. I've started to annotate with `# TODO: jaq` for things that needs a revisit.

**Comment by wader** [comment]
Oh seems eval/1 now works, not sure what fixed it, and also this 🥳 
```sh
$ jaq -L . 'include "jqjq"; eval("def f: 1,8; [f,f] | map(.+105) | implode")' <<< '{"a":123}'
"jqjq"
```

**Comment by wader** [comment]
Any plan to add `--args`, would make it possible to use the jqjq shell wrapper with jaq, `./jqjq --jq jaq ...`. But beware the semantics of `--args`, it's a bit confusing. Let me illustrate:

```sh
$ jq -n '$ARGS' --args a b c -- d
{
  "positional": [
    "a",
    "b",
    "c",
    "d"
  ],
  "named": {}
}

$ jq -n --args '$ARGS' a b c -- d
{
  "positional": [
    "a",
    "b",
    "c",
    "d"
  ],
  "named": {}
}
```

So skips `--` and the positional are the ones after the first non-option (the filter).

**Comment by wader** [comment]
Found another bug i think, `"" | test("")` outputs false with jaq, was minified from `test("^\\s*$")` in the test parse code. I tried to debug the `matches` function but my rust is not good enough yet. Read the regex crate documentation about empty matches but didn't see that would make it skip them by default.

But with an ugly hack, that is now in the jaq brach, i managed to run the tests! but not sure i would trust them just yet :) i did notice the test result output look a bit wrong so needs some work.

```sh
$ jaq -n -L . -rRs 'include "jqjq"; . | jqjq(["", "--run-tests"]; {})' < jqjq.test
<cut>
OK: line 1230: null | eval("eval(\"empty\")") -> null
DIFF: line 1238: null | (.a | eval(".a, .b.c")) = 123 -> {"a":{"a":123,"b":{"c":123}}}
  Expected: {"a":{"a":123,"b":{"c":123}}}
    Actual: 123
OK: line 1244: null | try eval("{a b}") catch "failed" -> "failed"
OK: line 1249: null | try eval(". as {$a $b} | .") catch "failed" -> "failed"
191 of 253 tests passed
null
```

**Comment by 01mf02** [comment]
Hi @wader, amazing work! Thanks so much for it.

> Doh, yes you correct, i build without `--release`. Now i've updated to latest main and built with release target, much faster 👍

It has happened to (nearly) everyone ... ;)

> When i read about regex-list it seem to focus on binary size, compile time and some less features (less unicode support?) over performance? but it was faster in this case also? ... oh they mean compile time as in time compiling a regex not rust build time? :)

I think that the statement in `regex-lite` refers to Rust build time. Still, I believe that by having a reduced feature set, the time to compile a regular expression is also reduced. This is probably not advertised by `regex-lite` because they assume that you are usually not compiling thousands of regular expressions. :) Which brings us to your next point ...

> BTW for regex heavy filters maybe a regex cache would be an interesting idea? there has been some talks in jq issues about various ways of doing, ex in [jqlang/jq#2856](https://github.com/jqlang/jq/issues/2856), maybe a LRU cache etc? i've also experimented a bit with how `/regex/` literals could work in jq, that would be neat and also simplify how to reuse a compiled regex.

I have already thought about this the last few days. And following your comment, I just implemented regex caching with an LRU cache. However, the result is underwhelming: Using `jaq -L . -n 'include "jqjq"; _builtins_src | lex'` as benchmark, we get 0.35 seconds without caching and 0.30 seconds with caching.
I also tried `jaq -n 'limit(100000; repeat("--")) | match("^-="; "").string'`, and here, we get 0.20 seconds without caching and 0.16 seconds with caching.

Given that the performance improvement is not that large, I will probably not integrate this into jaq.

For reference, here is the diff to reproduce the experiment:

~~~ diff
diff --git a/jaq-std/Cargo.toml b/jaq-std/Cargo.toml
index b9c3f84..db4d8da 100644
--- a/jaq-std/Cargo.toml
+++ b/jaq-std/Cargo.toml
@@ -29,6 +29,7 @@ libm = { version = "0.2.7", optional = true }
 aho-corasick = { version = "1.0", optional = true }
 base64 = { version = "0.22", optional = true }
 urlencoding = { version = "2.1.3", optional = true }
+lru = "0.12.5"
 
 [dev-dependencies]
 jaq-json = { path = "../jaq-json" }
diff --git a/jaq-std/src/regex.rs b/jaq-std/src/regex.rs
index 3378306..e7a9ee5 100644
--- a/jaq-std/src/regex.rs
+++ b/jaq-std/src/regex.rs
@@ -4,7 +4,7 @@ use alloc::string::{String, ToString};
 use alloc::vec::Vec;
 use regex_lite::{self as regex, Error, Regex, RegexBuilder};
 
-#[derive(Copy, Clone, Default)]
+#[derive(Copy, Clone, PartialEq, Eq, Default, Hash)]
 pub struct Flags {
     // global search
     g: bool,
@@ -65,6 +65,25 @@ impl Flags {
         let mut builder = RegexBuilder::new(re);
         self.impact(&mut builder).build()
     }
+
+    #[cfg(feature = "std")]
+    pub fn regex_cached(self, re: &str) -> Result<Regex, Error> {
+        use core::cell::RefCell;
+        use core::num::NonZeroUsize;
+        use lru::LruCache;
+
+        std::thread_local! {
+            static CACHE: RefCell<LruCache<(String, Flags), Result<Regex, Error>>> =
+                RefCell::new(LruCache::new(NonZeroUsize::new(64).unwrap()));
+        }
+
+        CACHE.with(|cache| {
+            cache
+                .borrow_mut()
+                .get_or_insert((re.to_string(), self), || self.regex(re))
+                .clone()
+        })
+    }
 }
 
 /// Mapping between byte and character indices.
~~~

**Comment by 01mf02** [comment]
> Without any changes it seem expressions like `1+2` just output `1`, but if i add `... | debug` or `... | .` to both it seem to work. Some tail or return optimisation going wrong? 🤔

Oh, oh, that looks indeed like tail recursion gone wrong. I'm looking into it, but so far, I was not able to reproduce it with a smaller example ...

**Comment by 01mf02** [comment]
> Any plan to add `--args`, would make it possible to use the jqjq shell wrapper with jaq, `./jqjq --jq jaq ...`.

That should actually not be that hard to do with [`trailing_var_arg`](https://docs.rs/clap/4.1.8/clap/builder/struct.Arg.html#method.trailing_var_arg).

**Comment by wader** [comment]
> Hi @wader, amazing work! Thanks so much for it.

Thank and same to you! also great that jqjq gets to be used for something useful!

> > Doh, yes you correct, i build without `--release`. Now i've updated to latest main and built with release target, much faster 👍
> 
> It has happened to (nearly) everyone ... ;)

👯 

> Given that the performance improvement is not that large, I will probably not integrate this into jaq.

Aha that's a bummer and a bit unintuitive 🤔 but maybe compiling a small regex:s is quite fast. Maybe something to revisit in the future if other things gets more optimised?

**Comment by 01mf02** [comment]
> Aha that's a bummer and a bit unintuitive 🤔 but maybe compiling a small regex:s is quite fast. Maybe something to revisit in the future if other things gets more optimised?

Yes, definitely. That's also why I posted the code here. ;)

**Comment by 01mf02** [comment]
I am trying to reproduce the (tail recursion?) bug you found. My current minimal program is this:

~~~ jq
def _e($query):
  $query as { $num, $op } |
  if $num then $num
  else
    $query as {$left, $right} |
    # TODO: does not work without ... | debug hmm
    def _l: _e($left) ;# | debug({l: .});
    def _r: _e($right) ;# | debug({r: .});
    _l + _r
  end
;
_e({"op":"+","left":{"num":1},"right":{"num":2}})
~~~

By the way, instead of writing `if .term == null then term = {} end`, you may use `.term //= {}`. :)

**Comment by wader** [comment]
> I am trying to reproduce the (tail recursion?) bug you found. My current minimal program is this:
> 
> ```jq
> def _e($query):
>   $query as { $num, $op } |
>   if $num then $num
>   else
>     $query as {$left, $right} |
>     # TODO: does not work without ... | debug hmm
>     def _l: _e($left) ;# | debug({l: .});
>     def _r: _e($right) ;# | debug({r: .});
>     _l + _r
>   end
> ;
> _e({"op":"+","left":{"num":1},"right":{"num":2}})
> ```

Nice, i was about to try minimize down a reproduction case also.

> By the way, instead of writing `if .term == null then .term = {} end`, you may use `.term //= {}`. :)

Ah true :) for jqjq i've tried to be quite verbose codewise for my own and other ppls sanity :) but maybe `//=`would be ok in some places.

**Comment by 01mf02** [comment]
Just a quick update: My current minimal example to reproduce the tail recursion bug is:

~~~ jq
def f:
  if . then . else
    def g: f;
    2 | g + g
  end;
f
~~~

This yields 2, whereas it should yield 4. The problem, indeed, is that the call of `f` in `g` is considered tail-recursive, but the call to `g` (in `g + g`) is not considered tail-recursive, so the return value of `g` is directly returned. This is clearly wrong, but I'll probably need some time to work out how to tackle this problem correctly. Anecdotally, I would say that tail recursion has been responsible for about 100% of the bugs in the compiler ... :/

**Comment by wader** [comment]
> Just a quick update: My current minimal example to reproduce the tail recursion bug is:
> 
> ```jq
> def f:
>   if . then . else
>     def g: f;
>     2 | g + g
>   end;
> f
> ```
> 
> This yields 2, whereas it should yield 4. The problem, indeed, is that the call of `f` in `g` is considered tail-recursive, but the call to `g` (in `g + g`) is not considered tail-recursive, so the return value of `g` is directly returned. This is clearly wrong, but I'll probably need some time to work out how to tackle this problem correctly.

I see, but nice work!

> Anecdotally, I would say that tail recursion has been responsible for about 100% of the bugs in the compiler ... :/

Yeap i've spent some time reading and scratching my head over trying to improve gojq's tail call code but didn't get far :)

**Comment by wader** [comment]
@01mf02 is there some quick hack to disable TCO for jaq? messing around with jqjq again but not sure if i'm running into same issue

**Comment by 01mf02** [comment]
@wader, sorry for the late reply! In the last days, I have worked quite hard on putting TCO on a more solid basis. I did not get around to test jqjq with it yet, but if you wish to: https://github.com/01mf02/jaq/pull/223.

**Comment by wader** [comment]
> @wader, sorry for the late reply! In the last days, I have worked quite hard on putting TCO on a more solid basis. I did not get around to test jqjq with it yet, but if you wish to: #223.

Nice, did a quick check `204 of 253 tests passed` 👍 

Skimming the failures the most seems to be about missing `setpath`, assignment (maybe also setpath somehow?), `[] | add`-difference, `195.0 % 16` difference, and probably a bunch because of null index that needs to be fixed.

```sh
$ jaq -n -L . -rRs 'include "jqjq"; . | jqjq(["", "--run-tests"]; {})' < jqjq.test | grep -v OK:
ERROR: line 14: null | 123, 1.23, 0.123, .123, .123e1, 1.23e1, 2E3, 1.23e+1, 1.23E-2 -> 123,1.23,0.123,0.123,1.23,12.3,2000,12.3,0.0123
  cannot parse .123 as JSON: value expected
ERROR: line 40: "b" | "aaa \("a" + . + "c") ccc" -> "aaa abc ccc"
  cannot use null as iterable (array or object)
ERROR: line 52: null | "aaa \(1,2,3) bbb" -> "aaa 1 bbb","aaa 2 bbb","aaa 3 bbb"
  cannot use null as iterable (array or object)
ERROR: line 95: "abcäሴ💩" | @uri "aaa \(.) bbb \(.)" -> "aaa abc%C3%A4%E1%88%B4%F0%9F%92%A9 bbb abc%C3%A4%E1%88%B4%F0%9F%92%A9"
  cannot calculate 195.0 % 16
ERROR: line 131: {"a":1} | {a} -> {"a":1}
  cannot use null as iterable (array or object)
ERROR: line 135: {"a":1} | {"a"} -> {"a":1}
  cannot use null as iterable (array or object)
ERROR: line 139: null | 1 as $a | {$a} -> {"a":1}
  cannot use null as iterable (array or object)
ERROR: line 147: null | {a: 1, b: 2} | 3 as $c | {a, "b", $c, ("d"): 4} -> {"a":1,"b":2,"c":3,"d":4}
  cannot use null as iterable (array or object)
ERROR: line 489: {"a":1,"b":2} | del(.a) -> {"b":2}
  undefined function delpaths/1
ERROR: line 493: [1,2,3] | del(.[1]) -> [1,3]
  undefined function delpaths/1
ERROR: line 497: [1,2,3] | del(.[1],.[1]) -> [1,3]
  undefined function delpaths/1
ERROR: line 527: null | .a = 123 -> {"a":123}
  cannot use null as iterable (array or object)
ERROR: line 531: null | (.a,.b) = 123 -> {"a":123,"b":123}
  cannot use null as iterable (array or object)
ERROR: line 536: null | .a = 123 | .a |= . + 1 -> {"a":124}
  cannot use null as iterable (array or object)
ERROR: line 540: null | .a = 123 | (.a,.b) |= . + 3 -> {"a":126,"b":3}
  cannot use null as iterable (array or object)
ERROR: line 546: null | .a = 123 | .a += 3 -> {"a":126}
  cannot use null as iterable (array or object)
ERROR: line 550: null | .a = 123 | (.a,.b) += 3 -> {"a":126,"b":3}
  cannot use null as iterable (array or object)
ERROR: line 555: null | .a = 3 | .a += .a + 3 -> {"a":9}
  cannot use null as iterable (array or object)
ERROR: line 561: null | .a = 123 | .a -= 3 -> {"a":120}
  cannot use null as iterable (array or object)
ERROR: line 565: null | .a = 123 | .b = 0 | (.a,.b) -= 3 -> {"a":120,"b":-3}
  cannot use null as iterable (array or object)
ERROR: line 571: null | .a = 123 | .a *= 3 -> {"a":369}
  cannot use null as iterable (array or object)
ERROR: line 575: null | .a = 123 | .b = 1 | (.a,.b) *= 3 -> {"a":369,"b":3}
  cannot use null as iterable (array or object)
ERROR: line 581: null | .a = 120 | .a /= 3 -> {"a":40}
  cannot use null as iterable (array or object)
ERROR: line 585: null | .a = 120 | .b = 3 | (.a,.b) /= 3 -> {"a":40,"b":1}
  cannot use null as iterable (array or object)
ERROR: line 591: null | .a = 123 | .a %= 4 -> {"a":3}
  cannot use null as iterable (array or object)
ERROR: line 595: null | .a = 123 | .b = 4 | (.a,.b) %= 4 -> {"a":3,"b":0}
  cannot use null as iterable (array or object)
DIFF: line 697: null | try (1+"a") catch . -> "number (1) and string (\"a\") cannot be added"
  Expected: "number (1) and string (\"a\") cannot be added"
    Actual: "cannot calculate 1 + \"a\""
DIFF: line 707: null | error(123)? // 456 -> 456
  Expected: 456
    Actual:
ERROR: line 811: [[],[{"a":1}],[{"a":1},{"a":2},{"a":1}]] | map(group_by(.a)) -> [[],[[{"a":1}]],[[{"a":1},{"a":1}],[{"a":2}]]]
  undefined function setpath/2
ERROR: line 815: [{"a":1,"b":2},{"a":2},{"a":1,"b":2}] | group_by(.a, .b) -> [[{"a":1,"b":2},{"a":1,"b":2}],[{"a":2}]]
  undefined function setpath/2
ERROR: line 819: [[],[{"a":1}],[{"a":1},{"a":2},{"a":1}]] | map(group_by(.a)) -> [[],[[{"a":1}]],[[{"a":1},{"a":1}],[{"a":2}]]]
  undefined function setpath/2
ERROR: line 823: [{"a":1,"b":2},{"a":2},{"a":1,"b":2}] | unique_by(.a, .b) -> [{"a":1,"b":2},{"a":2}]
  undefined function setpath/2
ERROR: line 827: [[],[1],[3,2,2],[3,1,1,2]] | map(unique) -> [[],[1],[2,3],[1,2,3]]
  undefined function setpath/2
ERROR: line 863: [{"key":"a","value":1},{"Key":"b","Value":2},{"Name":"c","value":false}] | from_entries -> {"a":1,"b":2,"c":false}
  undefined function setpath/2
ERROR: line 867: {"a":1,"b":2} | with_entries(.key |= "_"+. | .value += 100) -> {"_a":101,"_b":102}
  undefined function setpath/2
ERROR: line 916: 0 | [acos,acosh,asin,asinh,atan,atanh,cbrt,ceil,cos,cosh,erf,erfc,exp,exp10,exp2,expm1,fabs,floor,gamma,j0,j1,lgamma,log,log10,log1p,log2,logb,nearbyint,rint,round,significand,sin,sinh,sqrt,tan,tanh,tgamma,trunc,y0,y1] | map(type) -> ["number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number","number"]
  undefined function gamma/0
ERROR: line 968: {"a":{"c":1},"b":{"c":2}} | .[].c = 123 -> {"a":{"c":123},"b":{"c":123}}
  undefined function setpath/2
ERROR: line 983: [[1,2],[3,4]] | .[][] += 1 -> [[2,3],[4,5]]
  undefined function setpath/2
DIFF: line 998: null | empty // empty // 123 -> 123
  Expected: 123
    Actual:
DIFF: line 1002: null | . // . // 123 -> 123
  Expected: 123
    Actual:
DIFF: line 1006: null | (1,null,2,3) // 123 -> 1,2,3
  Expected: 1,2,3
    Actual: 1
DIFF: line 1012: null | null // (1,null,2,3) -> 1,null,2,3
  Expected: 1,null,2,3
    Actual:
DIFF: line 1019: null | null // 1,2 // 123 -> 1,2
  Expected: 1,2
    Actual: 2
DIFF: line 1029: null | try (empty // 1,error("a"),2) catch . -> 1,"a"
  Expected: 1,"a"
    Actual: "a"
ERROR: line 1035: {"foo":0} | (.foo,.foo,.foo) |= .+1 -> {"foo":3}
  undefined function setpath/2
ERROR: line 1078: [1,2,3] | bsearch(0,2,4) -> -1,1,-4
  undefined function setpath/2
ERROR: line 1159: {} | [path(.a, .b.c)] -> [["a"],["b","c"]]
  cannot use null as iterable (array or object)
ERROR: line 1183: [[],[1],[3,2,2],[3,1,1,2]] | map(group) -> [[],[[1]],[[2,2],[3]],[[1,1],[2],[3]]]
  undefined function setpath/2
ERROR: line 1238: null | (.a | eval(".a, .b.c")) = 123 -> {"a":{"a":123,"b":{"c":123}}}
  cannot use null as iterable (array or object)
204 of 253 tests passed
```

**Comment by wader** [comment]
A forgot, also tried with out `.. | .`-TCO-workaround and it seems to give same result

**Comment by 01mf02** [comment]
> A forgot, also tried with out `.. | .`-TCO-workaround and it seems to give same result

You mean that when you remove the `| .` part, you get the same result as with `| .`? I tested it just now, and this seems to be the case.

Now onward for `--args`: https://users.rust-lang.org/t/collecting-all-arguments-after-flag-with-clap/120674

**Comment by 01mf02** [comment]
> Nice, did a quick check `204 of 253 tests passed` 👍

Amazing! Thanks for testing the tests. :)

**Comment by wader** [comment]
> You mean that when you remove the `| .` part, you get the same result as with `| .`? I tested it just now, and this seems to be the case.

Yes exactly, so your changes seems to have fixed it.

> Now onward for `--args`: https://users.rust-lang.org/t/collecting-all-arguments-after-flag-with-clap/120674

👍 we (me and @thaliaarchi) wish to also in the future make the file `jqjq.jq` executable and remove the shell wrapper. But that requires the jq implementation to support comment escaping (from https://github.com/jqlang/jq/wiki/FAQ):
```sh
#!/bin/sh --
# this next line is ignored by jq, which otherwise does not continue comments \
exec jq -nef "$0" "$@"
# jq code follows
true
```

This works with jq and gojq (but jq 1.7.1 has a problem with `include`-ing such a file, fix in master).

**Comment by wader** [comment]
Created a draft PR for jqjq to track stuff https://github.com/wader/jqjq/pull/20

**Comment by 01mf02** [comment]
> 👍 we (me and @thaliaarchi) wish to also in the future make the file `jqjq.jq` executable and remove the shell wrapper. But that requires the jq implementation to support comment escaping (from https://github.com/jqlang/jq/wiki/FAQ):

This is now supported in jaq 0ef6ea3.

**Comment by 01mf02** [comment]
> Any plan to add `--args`, would make it possible to use the jqjq shell wrapper with jaq, `./jqjq --jq jaq ...`. But beware the semantics of `--args`, it's a bit confusing.

I have now implemented some support for `--args` #224. In jqjq, it will be necessary to move the 'include ...' part before `--args` to make it work.

**Comment by wader** [comment]
With latest main and small change to current jqjq jaq branch to add `-r` when using `-j`, is implicitly added with jq, not sure if bug or intentional for jaq? i can now do this:
```
$ ./jqjq --jq jaq -n --arg a 1 '"abc"'
"abc"
$ ./jqjq --jq jaq -n -r --arg a 1 '"abc"'
abc
$ ./jqjq --jq jaq -n --arg a 1 '123'
123
```

**Comment by wader** [comment]
btw i fixed some stuff in https://github.com/wader/jq-lsp so now it works much better with jaq's builtin jq files, and using this `.jq-lsp.jq` file it will also stop complaining about some missing functions:

```jq
def matches($re; $flags): empty;
def split_($re; $flags): empty;
def escape_csv: empty;
def escape_tsv: empty;
def escape_sh: empty;
def escape_html: empty;
def encode_uri: empty;
def encode_base64: empty;
def decode_base64: empty;
def ilogb: empty;
def isnumber: empty;
```

**Comment by 01mf02** [comment]
> With latest main and small change to current jqjq jaq branch to add `-r` when using `-j`, is implicitly added with jq, not sure if bug or intentional for jaq? i can now do this:

I'm afraid I do not understand what you want to say about `-r` and `-j`.
In any case, I re-cloned your repository (got a warning about "divergent branches"), then ran the commands you gave from the "jaq" branch.

~~~
$ ./jqjq --jq ../target/release/jaq -n --arg a 1 '"abc"'
"\u001b[0;32m""\"abc\"""\u001b[0m""\n"
$ ./jqjq --jq jq -n --arg a 1 '"abc"'
"abc"
~~~

**Comment by wader** [comment]
Sorry should have been more clear, with jq if you do -j it implies -r i think so will use raw output mode.

**Comment by wader** [comment]
Tracked down and minimized another difference/bug:
```
$ jq -n 'foreach (1,2,3) as $v (null; select($v != 2); $v)'
1
3
$ jaq -n 'foreach (1,2,3) as $v (null; select($v != 2); $v)'
1
```
but empty from project seems to behave the same
```sh
$ jq -n 'foreach (1,2,3) as $v (null; .; select($v != 2) | $v)'
1
3
$ jaq -n 'foreach (1,2,3) as $v (null; .; select($v != 2) | $v)'
1
3
```

**Comment by 01mf02** [comment]
> Sorry should have been more clear, with jq if you do -j it implies -r i think so will use raw output mode.

This seems to trip up more people than I expected, see #217. I am starting to wonder whether it is such a good idea in jaq to diverge from jq on this point ... What do you think?

**Comment by 01mf02** [comment]
> Tracked down and minimized another difference/bug:
> 
> ```
> $ jq -n 'foreach (1,2,3) as $v (null; select($v != 2); $v)'
> 1
> 3
> $ jaq -n 'foreach (1,2,3) as $v (null; select($v != 2); $v)'
> 1
> ```

Here, I'd argue that this is actually a bug in jq. The jq manual states that `foreach .[] as $item (0; . + $item; [$item, . * 2])` is "similar to":

~~~ jq
0 | 1 as $item | . + $item | [$item, . * 2],
    2 as $item | . + $item | [$item, . * 2],
    3 as $item | . + $item | [$item, . * 2]
~~~

So in that sense, we should be able to infer that `foreach (1,2,3) as $v (null; select($v != 2); $v)` is similar to:

~~~ jq
null | 1 as $item | select($item != 2) | $item,
       2 as $item | select($item != 2) | $item,
       3 as $item | select($item != 2) | $item
~~~

What does it yield? Just 1.

**Comment by wader** [comment]
> > Sorry should have been more clear, with jq if you do -j it implies -r i think so will use raw output mode.
> 
> This seems to trip up more people than I expected, see #217. I am starting to wonder whether it is such a good idea in jaq to diverge from jq on this point ... What do you think?

Yes maybe good idea. I think the only difference will be for strings? trying to think when someone would want non-separated quoted strings? 🤔

**Comment by wader** [comment]
To summarise current status and test failures:
- null index (including via destructing), some ugly workarounds in jqjq atm that needs to be made nicer
- A bunch fail on missing path functions like `setpath`will be implement in jaq? i guess jqqj could try reimplement them but maybe it's also good that jqjq exercises lots of different features?
- div/rem difference with float and integer, maybe mix of fixes in jaq and jqjq?
- `implode`  and float difference, probably fix in jqjq
- jqjq currently rely on null assignment,`null | .a = 1` => `{a:1}`, will have to be workaround around in jqjq i guess
- `-j` not implying `-r` difference
- `[] | join("") -> null` difference, jaq support joining other types than strings. workaround in jqjq atm

**Comment by wader** [comment]
> The jq's behavior of foreach hasn't changed since 1.5, and it's not a bug. The implementation in gojq was very straightforward and the behavior feels natural. The code in the jq manual lacks three parentheses.

True, but i also noticed that adding paratheses does not really describe the current behaviour:

```jq
foreach (1,2,3) as $item (0; .+1; {$item, dot: .})
# outputs
{"item":1,"dot":1}
{"item":2,"dot":2}
{"item":3,"dot":3}
```
While
```jq
0 | (1 as $item | .+1 | {$item, dot: .}),
    (2 as $item | .+1 | {$item, dot: .}),
    (3 as $item | .+1 | {$item, dot: .})
# outputs
{"item":1,"dot":1}
{"item":2,"dot":1}
{"item":3,"dot":1}
```

Also noticed that when adding `select($item != 2)` the behaviour changes, confused 🤔 
```jq
foreach (1,2,3) as $item (0; .+1 | select($item != 2); {$item, dot: .})
# outputs
{"item":1,"dot":1}
{"item":3,"dot":1}
```

**Comment by 01mf02** [comment]
> ```jq
> 0 | (1 as $item | .+1 | {$item, dot: .}),
>     (2 as $item | .+1 | {$item, dot: .}),
>     (3 as $item | .+1 | {$item, dot: .})
> # outputs
> {"item":1,"dot":1}
> {"item":2,"dot":1}
> {"item":3,"dot":1}
> ```

Watch out --- this actually pipes the initial accumulator (0) to all iterations. That's why `dot` is `1` for every output.

> Makes sense 👍 will figure some fix for jqjq to not rely on this. For some reason my head incorrectly parses this as 😄

Due to the precedences of `|` and `,` it actually parses to:

~~~ jq
null | 1 as $item | select($item != 2) | ($item,
       2 as $item | select($item != 2) | ($item,
       3 as $item | select($item != 2) | ($item)))
~~~

> The jq's behavior of foreach hasn't changed since 1.5, and it's not a bug.

What I was trying to say is: Either there is a bug in jq, or the jq manual does not actually specify the behaviour of jq.
It would be nice if jq and the jq manual could agree. But currently, they do not.

> The code in the jq manual lacks three parentheses.

Can you give us a corrected version of that code that corresponds to jq's behaviour?

**Comment by itchyny** [comment]
Sorry, I was confused too. The foreach behavior cannot be expressed like the pseudo code written in jq manual. It was me that wrote the part to make it easier to understand the syntax for users, but don't trust it for now. The expanded code is correct for some kind of expressions, maybe that doesn't include selections, but it doesn't explain [the backtracking behavior](https://github.com/jqlang/jq/blob/jq-1.7.1/src/compile.c#L882-L885) of the syntax. I also noticed the difference between jq and gojq, but not sure which behavior is more reasonable to formalize.

**Comment by 01mf02** [comment]
> The foreach behavior cannot be expressed like the pseudo code written in jq manual. It was me that wrote the part to make it easier to understand the syntax for users, but don't trust it for now.

@itchyny, I really like your pseudo code, and I think that it would be great if jq/gojq could actually simply do exactly what your pseudo code does. Right now, I think that it is quite hard to describe and understand the precise behaviour of jq's `foreach`, and bringing it closer to your pseudo code would be a win, IMHO. It should also not be that hard to implement this in a performant way; at least jaq has proven that it is possible.

**Comment by itchyny** [comment]
Yes, we can debate this topic, but we might come to conclusion that too late to change.

**Comment by 01mf02** [comment]
Alternatively, as some kind of middle ground, how about desugaring `foreach (1, 2, 3) as $v (INIT; UPDATE; EXTRACT)` to something like:

~~~ jq
def UPDATE_: label $x | (UPDATE | ., break $x), null;
INIT |
1 as $v | UPDATE_ | EXTRACT,
2 as $v | UPDATE_ | EXTRACT,
3 as $v | UPDATE_ | EXTRACT
~~~

This would be very close to the current jq behaviour, with the only difference being that if `UPDATE` yields no output, `EXTRACT` would be called with input `null`, whereas currently, `EXTRACT` would not be called in such a case.
I think that both jq/gojq could be adapted with minimal effort to this semantics, and that way, the documentation could accurately describe the behaviour of `foreach`. Backwards compatibility would only be hurt for cases where `UPDATE` yields multiple (or no) outputs, but given that the documentation does not actually document that case at all, it might be worth it.

**Comment by wader** [comment]
> > A bunch fail on missing path functions like `setpath`will be implement in jaq? i guess jqqj could try reimplement them but maybe it's also good that jqjq exercises lots of different features?
> 
> The problem is the `path/1` filter. Implementing this filter in jaq would require adding significant amounts of code in the core library of jaq, duplicating a lot of functionality. Given that jq uses `path/1` mostly for updates, which jaq implements differently, I am currently not motivated to implement this. What you could do, would be to provide some file like `polyfill-jaq.jq`:
> 
> ```jq
> def path(f): error("path not supported in jaq");
> def getpath(f): error("getpath not supported in jaq");
> def setpath(f): error("setpath not supported in jaq");
> ...
> ```
> 
> Then, to run jqjq, you could set `$POLYFILL` to `polyfill-jaq.jq` when jaq is used, and run something like `$JQ <(cat $POLYFILL jqjq.jq)`. That way, you can keep using `path`, `getpath` etc. and have everything working nicely in jq and gojq, while jaq will also be able to run jqjq, just failing on update operations.
> 
> Alternatively, if you are more adventurous, you could also try to implement `path` in jqjq yourself. :) But, to be honest, I wouldn't do it in your place, as I'd expect that to take much longer, and I'm not sure that it's worth it.

I see, and yes the path functions is quite advanced jq usage but `path/1` do allow for implement some quite interesting things. I had a quit look at jqjq and it seems like `path/1` is not actually used more than in the builtins where it's jqjq's own `path/1`. `setpath`/`getpath` is "passpthru" and used to implement assignment just like jq, `delpaths` is not used by jqjq itself, only passed thru.

**Comment by wader** [comment]
Not sure if it's what we want to do but i did my own, probably buggy, implementations of `setpath/1` and `getpath/1` and fixed a bunch of other null index cases. I've pushed the changes to the jaq branch if you want to have a look, now time for sleep!
```sh
$ ./jqjq --jq jaq --run-tests < jqjq.test | grep -v OK:
ERROR: line 487: {"a":1,"b":2} | del(.a) -> {"b":2}
  undefined function delpaths/1
ERROR: line 491: [1,2,3] | del(.[1]) -> [1,3]
  undefined function delpaths/1
ERROR: line 495: [1,2,3] | del(.[1],.[1]) -> [1,3]
  undefined function delpaths/1
250 of 253 tests passed
null
```

(`delpaths` i would guess is quite tricky to implement in jq? 🤔)

**Comment by wader** [comment]
@01mf02 would it be weird for jaq to implement `setpath/2`/`getpath/1`, should be easyish?, but not `path/1`?

I played around a bit yesterday with implementing `delpaths/1` in jq but no success yet. But i did find https://github.com/01mf02/jaq/issues/235 :) i was trying to do a variant that "boxes" all leaf values as `[.]`, use setpath to set tombstones, like null for each deleted path, recurse and filter out all tombstones, reverse the boxing. Turn out tricky to reverse it hmm

**Comment by 01mf02** [comment]
> @01mf02 would it be weird for jaq to implement `setpath/2`/`getpath/1`, should be easyish?, but not `path/1`?

`setpath/2` and `getpath/1` should both be easy to implement by definition. However, `path/1` is impossible to define by definition. It could be done in the core of jaq, but it is magnitudes more complex than `setpath` and `getpath`.
Would it help jqjq to have `setpath`/`getpath`, but not `path`?

**Comment by 01mf02** [comment]
> Not sure if it's what we want to do but i did my own, probably buggy, implementations of `setpath/1` and `getpath/1` and fixed a bunch of other null index cases. I've pushed the changes to the jaq branch if you want to have a look, now time for sleep!

Wow, I am extremely impressed by your work! 250/253 tests passed! I did not expect that in my wildest dreams. (Still, I hope that you are able to sleep well after such a _tour de force_.)

I have a suggestion for your `jqjq` wrapper script: Instead of passing `--jq jaq`, how about calling it with `JQ=jaq ./jqjq ...`?
In your script, you would need to set `JQ` to some default value by (taken from [here](https://stackoverflow.com/a/28085062)):

~~~ bash
: "${JQ:=jq}"
~~~

Then just use $JQ in the rest of the script. This would make it unnecessary to require `jq` (as the wrapper script currently does).

I also admit that I am not able to completely understand you approach with `invoke_client_jqjq`. Especially the part

~~~
    , "-L", "\"$(dirname \"$(realpath \"${BASH_SOURCE[0]}\")\")\""
~~~

was where I tuned out. I found your previous approach (where you kept most of the logic in the wrapper script `jqjq` instead of in `jqjq.jq`) much more understandable. May I be so curious to ask you why you moved the calling logic to jq? (My guess would be: regular expression support.)

**Comment by 01mf02** [comment]
I've looked at your `getpath` and `setpath` implementations, and I did not notice any flaws. Expect for the fact that you do not handle slices, but I'm not sure whether you want to do that. :)

**Comment by 01mf02** [comment]
> I see, and yes the path functions is quite advanced jq usage but `path/1` do allow for implement some quite interesting things. I had a quit look at jqjq and it seems like `path/1` is not actually used more than in the builtins where it's jqjq's own `path/1`. `setpath`/`getpath` is "passpthru" and used to implement assignment just like jq, `delpaths` is not used by jqjq itself, only passed thru.

Oh, I see! You actually already implement `path/1` in jqjq! Then please forget everything I said about polyfills and so on. :)

**Comment by 01mf02** [comment]
Concerning `delpaths`: How about something like:

~~~ jq
def _delpath($p):
  ($p | length) as $len |
    if $len == 0 then error
  elif $len == 1 then .[$p[0]] |= empty
  else .[$p[0]] |= _delpath($p[1:])
  end;
def _delpaths($paths): reduce ($paths | sort | reverse)[] as $p (.; _delpath($p));
~~~

Test:

~~~ jq
[0, 1, 2] | _delpaths([[0], [2]])
~~~

This yields `[1]`.

**Comment by 01mf02** [comment]
I have found a shorter version of `_delpaths`:

~~~ jq
def _delpath($p): if $p | length == 0 then empty else .[$p[0]] |= _delpath($p[1:]) end;
def _delpaths($paths): reduce ($paths | sort | reverse)[] as $p (.; _delpath($p));
~~~

This also handles empty paths (`[]`). However, note that when running this in jaq, `_delpaths([[]])` returns `empty`, whereas jq returns `null`. This is due to the difference in `reduce` semantics. So `_delpaths` probably needs to be adapted a bit if this should yield the same outputs for both jq and jaq.

Anyway, this is all quite approximative; the devil is in the details. But it should give you a good start nonetheless.

**Comment by 01mf02** [comment]
One last change for this week (hopefully): The following code now only deletes if the path is valid:

~~~ jq
def _delpath($p):
    if $p | length == 0 then empty
  elif has($p[0]) then .[$p[0]] |= _delpath($p[1:])
  else . end;

def _delpaths($paths): reduce ($paths | sort | reverse)[] as $p (.; _delpath($p));
~~~

This makes the following work in both jq and jaq, where jaq would previously fail with an error:

~~~ jq
[0, 1, 2] | _delpaths([[2], [2]])
~~~

This yields [0, 1].

**Comment by wader** [comment]
> Wow, I am extremely impressed by your work! 250/253 tests passed! I did not expect that in my wildest dreams. (Still, I hope that you are able to sleep well after such a _tour de force_.)

Thanks! and no worries, had a good sleep :) i was expecting there would be more issues. But there is still some cleanup to do, i just shoggled along with hacks to get a feel how bad it was :) turns out not so bad. 

With your `delpaths/1` we are now at 253/253 🥳
```
$ ./jqjq --jq jaq --run-tests < jqjq.test | grep -v OK:
253 of 253 tests passed
```

I reformatted it a bit to match the rest of jqjq and also changed `sort` to `unique`, there was an issue with deleting the same path in an array more than one time, i luckily for some reason had a test case for that, not sure why.

> I have a suggestion for your `jqjq` wrapper script: Instead of passing `--jq jaq`, how about calling it with `JQ=jaq ./jqjq ...`? In your script, you would need to set `JQ` to some default value by (taken from [here](https://stackoverflow.com/a/28085062)):
> 
> ```shell
> : "${JQ:=jq}"
> ```
> 
> Then just use $JQ in the rest of the script. This would make it unnecessary to require `jq` (as the wrapper script currently does).

Aha yes that is a good idea 👍 

> I also admit that I am not able to completely understand you approach with `invoke_client_jqjq`. Especially the part
> 
> ```
>     , "-L", "\"$(dirname \"$(realpath \"${BASH_SOURCE[0]}\")\")\""
> ```
> 
> was where I tuned out. I found your previous approach (where you kept most of the logic in the wrapper script `jqjq` instead of in `jqjq.jq`) much more understandable. May I be so curious to ask you why you moved the calling logic to jq? (My guess would be: regular expression support.)

The idea as mostly that me and @thaliaarchi have had this idea/obsession to do as much in jq as possible, so in https://github.com/wader/jqjq/pull/19 is was changed so that jq code is now used to also figure out how to start the "host" jq. This also means that the argument parsing code can be shared between the two "stages".

In some future when all stable versions of jq, gojq and jaq supports comment escaping the idea is to remove the jqjq wrapper shell script and make ./jqjq.jq as an executable, making jqjq 100% jq :) (+0.01% sh for the exec line)

**Comment by wader** [comment]
> I've looked at your `getpath` and `setpath` implementations, and I did not notice any flaws. Expect for the fact that you do not handle slices, but I'm not sure whether you want to do that. :)

Yeah i have kind of suppressed that jq can have slices in paths for now :)

**Comment by wader** [comment]
> setpath/2 and getpath/1 should both be easy to implement by definition. However, path/1 is impossible to define by definition. It could be done in the core of jaq, but it is magnitudes more complex than setpath and getpath.
Would it help jqjq to have setpath/getpath, but not path?

Yes `setpath`, `getpath` and `delpaths` would be good to have, and i guess good for jaq to have also even if `path` is missing?

> Oh, I see! You actually already implement path/1 in jqjq! Then please forget everything I said about polyfills and so on. :)

Yeap! maybe it could be possible to make the host jq do the path tracking, but think a ran into some problems. Anyways handling path made it possible to do some quite nice things like this:

```sh
$ gojq -cn -L . 'include "jqjq"; {a:0, b:1} | eval(".a, .b") += 1'
{"a":1,"b":2}
```

wonder if that will work with jaq as host jq? i tried add my own `getpath` so try but got this:
```
$ jaq -cn -L . 'include "jqjq"; {a:0, b:1} | eval(".a, .b") += 1'
thread 'main' panicked at /Users/wader/src/jaq/jaq-core/src/filter.rs:369:17:
not implemented: updating with this operator is not supported yet
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

make jqjq's `eval/3` look like this if you want to try:
```jq
def eval($expr; $globals; $builtins_env):
  ( eval_ast(
      $expr | lex | parse;
      [];
      $builtins_env;
      builtin_undefined_func($globals; $builtins_env)
    ) as [$path, $value]
  | if $path | . == [] or . == [null] then $value
    else getpath($path)
    end
  );
```

**Comment by wader** [comment]
BTW maybe it would be nice if jaq 2.0 could be released with support for jqjq?

**Comment by thaliaarchi** [comment]
Lots of interesting work going on here, so I want to chime in, but I can't contribute code right now, since I need to focus on other things.

- I don't think the implementation of `--args` is correct.

  A couple of nuances that don't look handled with clap: It should not change the interpretation of positional arguments before it. Flags can appear after it, until the first `--`. `--jsonargs` can be intermixed (just like any other flag) and works the same, but with JSON values; thus, positional args need to be tagged with whether they're a string or JSON.

  clap may be the wrong abstraction. I'd write an arg parser to more closely match [jq's arg parsing](https://github.com/jqlang/jq/blob/master/src/main.c#L345-L532). As another example, `--long=value` is not supported (but I want to fix that upstream).

- I don't think jaq should diverge from jq on `-j`. It's been in jq as-is for a long time, since jq 1.5, and been documented as “Like `-r` but jq won't print a newline after each output” from its start.

  It would, however, be nice to have options to fill out the matrix of flags (but I think they should be new flags):

  - `--raw-output`: `RAW_OUTPUT`
  - `--raw-output0`: `RAW_OUTPUT | RAW_NO_LF | RAW_OUTPUT0`
  - `--join-output`: `RAW_OUTPUT | RAW_NO_LF`

- In jqjq, `invoke_client_jqjq` is called by the host jq constructs a bash command line to invoke the client jqjq. After the flags, here's what it does:

  - `-L "$(dirname "$(realpath "${BASH_SOURCE[0]}")")"` adds the directory of jqjq.jq to the search path, so it can be included (though that only works within a bash script currently, so will need to be fixed).
  - `'include "jqjq"; jqjq($ARGS.positional; $ENV)'` is the filter that runs jqjq.
  - Next it takes all files from `-f`, inputs, `--rawfile`, and `--slurpfile`, deduplicates them, and sends a `--rawfile` variable to the client with the contents of the file. Those variables are named with the pattern `file:\($filename)`, so they won't be valid identifiers and have to be accessed through `$ARGS.named`.

    Adding support for files was the main reason for the move to processing options all in jq. Handling it in bash would need to replicate the entire arg parsing logic, so it could find all places that require files. Instead in jq, both stages parse arguments the same and there is no mismatch or duplication.

  - Finally, the entire command line is passed after `--args --` for it to be re-parsed by the client jqjq.

**Comment by 01mf02** [comment]
> BTW maybe it would be nice if jaq 2.0 could be released with support for jqjq?

That would be my plan. Given that jqjq now passes all tests when run with jaq, what is there left to be done?

**Comment by 01mf02** [comment]
> > setpath/2 and getpath/1 should both be easy to implement by definition. However, path/1 is impossible to define by definition. It could be done in the core of jaq, but it is magnitudes more complex than setpath and getpath.
> > Would it help jqjq to have setpath/getpath, but not path?
> 
> Yes `setpath`, `getpath` and `delpaths` would be good to have, and i guess good for jaq to have also even if `path` is missing?

Hmmm ... I have to meditate over this a bit. I really do not like paths ...

> Yeap! maybe it could be possible to make the host jq do the path tracking, but think a ran into some problems. Anyways handling path made it possible to do some quite nice things like this:
> 
> ```shell
> $ gojq -cn -L . 'include "jqjq"; {a:0, b:1} | eval(".a, .b") += 1'
> {"a":1,"b":2}
> ```
> 
> wonder if that will work with jaq as host jq? i tried add my own `getpath` so try but got this:
> 
> ```
> $ jaq -cn -L . 'include "jqjq"; {a:0, b:1} | eval(".a, .b") += 1'
> thread 'main' panicked at /Users/wader/src/jaq/jaq-core/src/filter.rs:369:17:
> not implemented: updating with this operator is not supported yet
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

I suppose that you used your `_getpath` as `getpath` here. It is not surprising that this does not work, because `getpath` uses `reduce`, and AFAIK, in both `jq` and `jaq`, you cannot update with `reduce ... (...) |= f`.

**Comment by 01mf02** [comment]
> I reformatted it a bit to match the rest of jqjq and also changed `sort` to `unique`, there was an issue with deleting the same path in an array more than one time, i luckily for some reason had a test case for that, not sure why.

Is that really correct? Shouldn't you first `sort`, then `unique`? Without `sort`, `[0, 1, 2] | delpaths([[1], [0]])` will yield `[1]` instead of `[2]`, I think.

**Comment by thaliaarchi** [comment]
> Per chance, do you know of a CLI parser (preferably with `derive` support) that could handle this use case?

Nope. I've only ever used `clap`.

After I cleaned up some bugs in jq's option parser (in https://github.com/jqlang/jq/pull/3194), then ported it to jq (in https://github.com/wader/jqjq/pull/19), I started porting it to Rust, since I've wanted to make my own jq implementation for a while. A clone of jq's logic wouldn't be much work.

**Comment by wader** [comment]
> > BTW maybe it would be nice if jaq 2.0 could be released with support for jqjq?
> 
> That would be my plan. Given that jqjq now passes all tests when run with jaq, what is there left to be done?

Yeap i think we're in quite good shape and jqjq can probably adapt to what jaq 2.0 ends up supporting.

**Comment by wader** [comment]
I noticed that `--repl` behaves different with jaq, i'm guessing output line buffering difference. Haven't dugg deeper into it yet.

```sh
# the "%" is some zsh indicator about missing new line
$ jq -Rjn '"prompt> " | ., input'
prompt> 123
123%

$ jaq -Rjn '"prompt> " | ., input'
123
prompt> 123%
$ jaq -Rjn '"prompt>\n" | ., input'
prompt>
123
123%
```

**Comment by wader** [comment]
> > I reformatted it a bit to match the rest of jqjq and also changed `sort` to `unique`, there was an issue with deleting the same path in an array more than one time, i luckily for some reason had a test case for that, not sure why.
> 
> Is that really correct? Shouldn't you first `sort`, then `unique`? Without `sort`, `[0, 1, 2] | delpaths([[1], [0]])` will yield `[1]` instead of `[2]`, I think.

`unique` both removes dups and sorts so sorting should not be needed. From jq docs:

> The unique function takes as input an array and produces an array of the same elements, in sorted order, with duplicates removed.

**Comment by 01mf02** [comment]
> After I cleaned up some bugs in jq's option parser (in [jqlang/jq#3194](https://github.com/jqlang/jq/pull/3194)), then ported it to jq (in [wader/jqjq#19](https://github.com/wader/jqjq/pull/19)), I started porting it to Rust, since I've wanted to make my own jq implementation for a while. A clone of jq's logic wouldn't be much work. ([`OsStr::as_encoded_bytes`](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.as_encoded_bytes) would be useful.)

You are right. I have now rewritten the command-line argument parser, see #216.

**Comment by 01mf02** [comment]
> `unique` both removes dups and sorts so sorting should not be needed.

You are right; my brain was in Rust mode.

**Comment by 01mf02** [comment]
> I noticed that `--repl` behaves different with jaq, i'm guessing output line buffering difference. Haven't dugg deeper into it yet.
> 
> ```shell
> # the "%" is some zsh indicator about missing new line
> $ jq -Rjn '"prompt> " | ., input'
> prompt> 123
> 123%
> 
> $ jaq -Rjn '"prompt> " | ., input'
> 123
> prompt> 123%
> $ jaq -Rjn '"prompt>\n" | ., input'
> prompt>
> 123
> 123%
> ```

Your intuition is correct; replacing `print` with the following function in `jaq/src/main.rs`, your example works:

~~~ rust
fn print(writer: &mut impl Write, cli: &Cli, val: &Val) -> io::Result<()> {
    let f = |f: &mut Formatter| fmt_val_root(f, cli, val);
    write!(writer, "{}", FormatterFn(f))?;
    writer.flush()
}
~~~

This might be a good time to implement `--unbuffered` ... but not today. :)

**Comment by wader** [comment]
> > > setpath/2 and getpath/1 should both be easy to implement by definition. However, path/1 is impossible to define by 
> Hmmm ... I have to meditate over this a bit. I really do not like paths ...

👍 jqjq can always reimplement for now, and yes i can understand that it's a bit weird to have just some path related functions and not all.

> > $ jaq -cn -L . 'include "jqjq"; {a:0, b:1} | eval(".a, .b") += 1'
> > thread 'main' panicked at /Users/wader/src/jaq/jaq-core/src/filter.rs:369:17:
> > not implemented: updating with this operator is not supported yet
> > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > ```
> 
> I suppose that you used your `_getpath` as `getpath` here. It is not surprising that this does not work, because `getpath` uses `reduce`, and AFAIK, in both `jq` and `jaq`, you cannot update with `reduce ... (...) |= f`.

Ok i see, yeap used the reimplementations. But shouldn't it ideally cause some jq runtime error instead of panic?

**Comment by 01mf02** [comment]
> 👍 jqjq can always reimplement for now, and yes i can understand that it's a bit weird to have just some path related functions and not all.

Yes.

> But shouldn't it ideally cause some jq runtime error instead of panic?

Yes. Originally, I left the panic at that place because I wasn't sure whether it was possible to perform updates (`p |= f`) with certain filters (`p`) on the left-hand side. Now I'm rather sure that it is not possible in jaq's update model to have `label $x | ...` and `try ... catch ...` on the left-hand side of updates, whereas `reduce` and `foreach` could be possible in principle. However, because updating with `reduce` and `foreach` is not possible in jq neither, I have decided to do the same thing for all these filters, namely to return a path expression error, like jq. c41632d3c6401b457063394fe8f984bc72d7615b

**Comment by 01mf02** [comment]
> > I noticed that `--repl` behaves different with jaq, i'm guessing output line buffering difference. Haven't dugg deeper into it yet.

I corrected this in a51d8daba653c13da88a0e47ad5ea23fb4431608. Thanks again for bringing this up, and for all the other issues you reported. All your feedback has really helped me to improve the quality of jaq! I am very grateful for your help.

**Comment by 01mf02** [comment]
By the way, I am planning to release jaq 2.0 this week, probably tomorrow. I was thinking about mentioning that jaq can now run jqjq. Are there plans from your side to merge jqjq's jaq branch into its master branch, or should I give instructions in the release notes for jaq 2.0 to continue using jqjq's jaq branch?

**Comment by wader** [comment]
> > But shouldn't it ideally cause some jq runtime error instead of panic?
> 
> Yes. Originally, I left the panic at that place because I wasn't sure whether it was possible to perform updates (`p |= f`) with certain filters (`p`) on the left-hand side. Now I'm rather sure that it is not possible in jaq's update model to have `label $x | ...` and `try ... catch ...` on the left-hand side of updates, whereas `reduce` and `foreach` could be possible in principle. However, because updating with `reduce` and `foreach` is not possible in jq neither, I have decided to do the same thing for all these filters, namely to return a path expression error, like jq. [c41632d](https://github.com/01mf02/jaq/commit/c41632d3c6401b457063394fe8f984bc72d7615b)

👍 when messing with reduce/foreach for jqjq i did notice that it's possible for them to be valid path expressions but it might not be very useful:

```sh
$ jq -cn '{a:1} | path((foreach (1,2,3) as $v (.;.)).a)'
["a"]
["a"]
["a"]
$ jq -cn '{a:1} | (foreach (1,2,3) as $v (.;.)).a += 1'
{"a":4}

$ jq -cn '{a:1} | path((reduce (1,2,3) as $v (.;.)).a)'
["a"]
$ jq -cn '{a:1} | (reduce (1,2,3) as $v (.;.)).a += 1'
{"a":2}
```

**Comment by wader** [comment]
> By the way, I am planning to release jaq 2.0 this week, probably tomorrow. I was thinking about mentioning that jaq can now run jqjq. Are there plans from your side to merge jqjq's jaq branch into its master branch, or should I give instructions in the release notes for jaq 2.0 to continue using jqjq's jaq branch?

🥳 good you reminded me, yes hope i can get some time tonight to cleanup and merge things to master. it's mostly the null index workarounds i've been a bit reluctant about how to fix but i can probably come up with something.

**Comment by wader** [comment]
@01mf02 jqjq master should now work with jaq 2.0. there are still some TODOs left but i think they can wait

**Comment by 01mf02** [comment]
> 👍 when messing with reduce/foreach for jqjq i did notice that it's possible for them to be valid path expressions but it might not be very useful:
> 
> ```shell
> $ jq -cn '{a:1} | path((foreach (1,2,3) as $v (.;.)).a)'
> ["a"]
> ["a"]
> ["a"]
> $ jq -cn '{a:1} | (foreach (1,2,3) as $v (.;.)).a += 1'
> {"a":4}
> 
> $ jq -cn '{a:1} | path((reduce (1,2,3) as $v (.;.)).a)'
> ["a"]
> $ jq -cn '{a:1} | (reduce (1,2,3) as $v (.;.)).a += 1'
> {"a":2}
> ```

Oh, that's interesting! However, this seems to be quite limited; consider the following example:

~~~
$ jq -n '[[[2], 1], 0] | foreach (0, 0) as $x (.; .[$x]) |= .'
jq: error (at <unknown>): Invalid path expression near attempt to access element 0 of [[2],1]
~~~

**Comment by wader** [comment]
> Oh, that's interesting! However, this seems to be quite limited; consider the following example:
> 
> ```
> $ jq -n '[[[2], 1], 0] | foreach (0, 0) as $x (.; .[$x]) |= .'
> jq: error (at <unknown>): Invalid path expression near attempt to access element 0 of [[2],1]
> ```

Yeap agree, not sure how useful it is, but was a bit surprised why i wouldn't work as long as you make sure to only index into the input. Intentional or bug?

```sh
$ ./jqjq -cn '[[2], 1] | path(reduce (0, 0) as $x (.; .[$x]))'
[0,0]
$ ./jqjq -cn '[[2], 1] | (reduce (0, 0) as $x (.; .[$x])) |= 3'
[[3],1]
$ ./jqjq -cn '[[2], 1] | path(foreach (0, 0) as $x (.; .[$x]))'
[0]
[0,0]
# 🤔
$ ./jqjq -cn '[[2], 1] | (foreach (0, 0) as $x (.; .[$x])) = .'
[[[[2],1],1],1]
```

**Comment by 01mf02** [comment]
Oh, this works in jqjq! Just checked gojq:

~~~
$ bin/gojq-0.12.16 -n '[[[2], 1], 0] | foreach (0, 0) as $x (.; .[$x]) |= .'
gojq: invalid path against: array ([[2],1])
~~~

That means that to my knowledge, jqjq is the only jq implementation that handles nontrivial paths in `reduce`/`foreach`. Congratulations! 🥳

**Comment by 01mf02** [comment]
> @01mf02 jqjq master should now work with jaq 2.0. there are still some TODOs left but i think they can wait

Thanks a lot! I just tested it and it seems to work! (Not sure whether you know it, but the "fib.bf" example from the README seems to run out of memory for both jaq and jq.)

**Comment by wader** [comment]
> > @01mf02 jqjq master should now work with jaq 2.0. there are still some TODOs left but i think they can wait
> 
> Thanks a lot! I just tested it and it seems to work! (Not sure whether you know it, but the "fib.bf" example from the README seems to run out of memory for both jaq and jq.)

👍 did't know fib was OOM:ing but i know jqjq via jqjq does atm, i used to work before some try/catch change i think that for some reason caused lots of more memory usage, would be nice to make it work again.

**Comment by wader** [comment]
Planning on doing a release today? let me know if i can help out with something

**Comment by 01mf02** [comment]
> Planning on doing a release today? let me know if i can help out with something

I've just created the release. :) If you are motivated to do this, it would be definitely helpful to post the news about the release to some site where people could be interested about this release, e.g. [Hacker News](https://news.ycombinator.com). (I'm having a bit of trouble doing this kind of "publicity".) I'll write a post on Reddit later today --- stepping out of my comfort zone for this. 😬

**Comment by itchyny** [comment]
The path filter should emit error if the value at which indexing begins is different from the value on which the path filter was called. For example, in `[{"x":1}] | path(.[0] as $x | $x | .x)`, the indexing `.x` is called on `{"x":1}`, which is different from `[{"x":1}]`, so should be an error. If not, it emits the result `["x"]`, which is an invalid path for `[{"x":1}]`. The `foreach` example is like `[[[2], 1], 0] | (.[0] as $x | $x | .[0]) = 1`, which should be an error.

**Comment by wader** [comment]
> > Planning on doing a release today? let me know if i can help out with something
> 
> I've just created the release. :) If you are motivated to do this, it would be definitely helpful to post the news about the release to some site where people could be interested about this release, e.g. [Hacker News](https://news.ycombinator.com). (I'm having a bit of trouble doing this kind of "publicity".) I'll write a post on Reddit later today --- stepping out of my comfort zone for this. 😬

Will do, have already posted to jq discord and hacker news https://news.ycombinator.com/item?id=42254530 will post on mastodon and maybe some more places also. You should definitely write some posts about it! i'm very curious and hopeful that ppl will build stuff around the new data format agnostic support.

**Comment by wader** [comment]
> The path filter should emit error if the value at which indexing begins is different from the value on which the path filter was called. For example, in `[{"x":1}] | path(.[0] as $x | $x | .x)`, the indexing `.x` is called on `{"x":1}`, which is different from `[{"x":1}]`, so should be an error. If not, it emits the result `["x"]`, which is an invalid path for `[{"x":1}]`. The `foreach` example is like `[[[2], 1], 0] | (.[0] as $x | $x | .[0]) = 1`, which should be an error.

By "if the value at which indexing begins is different from the value on which the path filter was called" you mean the value that beings each foreach/reduce iteration? my thinking is that `path(reduce (0, 0) as $_ (.; .[0]))` is more or less ` path(.[0] | .[0])`.

**Comment by 01mf02** [comment]
> Will do, have already posted to jq discord and hacker news https://news.ycombinator.com/item?id=42254530 will post on mastodon and maybe some more places also. You should definitely write some posts about it! i'm very curious and hopeful that ppl will build stuff around the new data format agnostic support.

Thanks a lot for this!

**Comment by wader** [comment]
👍 also posted here https://lobste.rs/s/5lvjcn/jaq_2_0_released_with_lots_improvements

**Comment by wader** [comment]
Time to close or something left? there was some `reduce`/`foreach` fallout but i guess that could continue in the jq repo

**Comment by 01mf02** [comment]
> 👍 also posted here https://lobste.rs/s/5lvjcn/jaq_2_0_released_with_lots_improvements

Thanks a lot again for the publicity --- I highly appreciate it!

**Comment by 01mf02** [comment]
> Time to close or something left? there was some `reduce`/`foreach` fallout but i guess that could continue in the jq repo

Yes, I think we can close this. The `reduce`/`foreach` "fallout" should indeed be discussed in the jq repo --- it's a bit unfortunate that nobody has commented on this so far.

**Comment by 01mf02** [comment]
In any case, it has been a lot of fun working with you on this issue, @wader!

**Comment by wader** [comment]
> > Time to close or something left? there was some `reduce`/`foreach` fallout but i guess that could continue in the jq repo
> 
> Yes, I think we can close this. The `reduce`/`foreach` "fallout" should indeed be discussed in the jq repo --- it's a bit unfortunate that nobody has commented on this so far.

👍 yeap i saw your issue, feel a bit bad that i haven't spent much time thinking about it, hopefully this week!

**Comment by wader** [comment]
> In any case, it has been a lot of fun working with you on this issue, @wader!

Same! hope (and i'm quite sure) we can find some more stuff to collaborate on!
